### PR TITLE
Modified initial version of the data pipeline nodes

### PR DESCRIPTION
**Pull request by pjakovits**
## Short Description

This addition introduces the Data Pipeline models and a set of typical Data pipeline node types together with Ansible implementations.  

## Resolving Issue / Feature
Fixed the issues with the earlier pull request in regards to TOSCA components versions and accidentally modified images. 

Contributors: Chinmaya Dehury (Models and node types), Pelle Jakovits

### CODE REVIEW

File: artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca [hunk 1]
```diff
@@ -0,0 +1,8 @@
+tosca_definitions_version: tosca_simple_yaml_1_3
+
+artifact_types:
+  installPipeline:
+    metadata:
+      targetNamespace: tosca.artifacts
```
Comment (miwurster) [inline artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca]: Pls avoid using the `tosca.*` namesapce

File: artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca [hunk 2]
```diff
@@ -0,0 +1,8 @@
+tosca_definitions_version: tosca_simple_yaml_1_3
+
+artifact_types:
+  installPipeline:
```
Comment (miwurster) [inline artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca]: Please use `PascalCase` for type names.

File: capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca [hunk 3]
```diff
@@ -0,0 +1,9 @@
+tosca_definitions_version: tosca_simple_yaml_1_3
+
+capability_types:
+  ConnectToPipeline:
```
Comment (miwurster) [inline capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca]: The type names should always contain the respective namespace. See the other types in the particles repo.

File: capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca [hunk 4]
```diff
@@ -0,0 +1,9 @@
+tosca_definitions_version: tosca_simple_yaml_1_3
```
Comment (miwurster) [inline capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca]: It seems you are using a wrong filename. It should be `<EntityType>.tosca`. Please see the other types of the particles.

**Review by miwurster — CHANGES_REQUESTED** [review]
It seems you created a couple of types by hand - see review comments; these comments actually apply to all of your files. Please make sure that GMT (Eclipse Winery) can process the types, otherwise, we cannot guarantee to create a valid CSAR.

File: artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca (hunk 2 above)
Comment (pjakovits) [reply artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca]: Fixed this for installPipeline. Did not notice the same issue elsewhere.

File: artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca (hunk 1 above)
Comment (pjakovits) [reply artifacttypes/tosca.artifacts/installPipeline/installPipeline.tosca]: Moved it to radon.artifacts.datapipeline

File: capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca (hunk 4 above)
Comment (pjakovits) [reply capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca]: Some of the manual fixes to errors introduced in removal of versions caused this. Should be solved now.

File: capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca (hunk 3 above)
Comment (pjakovits) [reply capabilitytypes/radon.capabilities.datapipeline/ConnectToPipeline/ConnectToPipeline.tosca]: Solved this issue in 3 files. Did not find other occurrences.

**Comment by pjakovits** [comment]
Modified the types according to the raised issues. Made sure the example data pipeline template (renamed to DataPipelineExample) can be downloaded as CSAR and edited without issues.

**Review by chinmaya-dehury — CHANGES_REQUESTED** [review]
For `nodetypes/radon.nodes.datapipeline.destination/PublishDataEndPoint/NodeType.tosca` it is not necessary to keep two identical capability types. So we may remove `connectToPipeline` and just keep `ConnectToPipeline`.

**Comment by miwurster** [comment]
One more thing, could you please make sure you try to process your new types with the latest version Docker image of Winery (`docker pull opentosca/radon-gmt:latest`, see [here](https://eclipse-winery.readthedocs.io/en/latest/user/getting-started.html#launching-with-docker))?

**Comment by pjakovits** [comment]
Yes, It is working on the latest version. 
Command:
`docker pull opentosca/radon-gmt:latest`
did not do to any updates to the image. 

Command I have been using to test it is: 
`docker run -it -p 8084:8080   --name wineryp -d     -e WINERY_FEATURE_RADON=true   -e WINERY_REPOSITORY_PROVIDER=yaml   -e WINERY_REPOSITORY_URL=https://github.com/pjakovits/radon-particles   opentosca/radon-gmt`


Can edit the example data pipeline service template with graphical editor and also download it as CSAR.

**Comment by miwurster** [comment]
Very nice. There is one hidden feature in Winery to validate if all entities in a repository are valid. If you start Winery you could basically to `POST` call to `http://localhost:8080/winery/admin/repository/touch`. What it does it loads every entity inside the repository and saves it again. This guarantees that Winery is able to process everything correctly.

Further, you could start Winery with an external repository configuration as described here: <https://eclipse-winery.readthedocs.io/en/latest/user/getting-started.html#use-a-custom-tosca-model-repository>. With `-v <path_on_your_host>:/var/repository` you could mount a manually cloned Git repository into the Winery container.

**Comment by pjakovits** [comment]
The command:
curl -d "" http://IP:8084/winery/admin/repository/touch 
returned empty answer when sending a empty string as data.

**Comment by pjakovits** [comment]
I have also tried to use the
`-v <path_on_your_host>:/var/repository`
option, but ran into some issues with file permissions. Perhaps I was not supposed to modify files from outside the container.

**Comment by miwurster** [comment]
Empty `<200>` response is fine and expected.

Yeah, I also experienced some permission issues when working on Linux. What OS are u using?

**Review by chinmaya-dehury — APPROVED** [review]
For me this seems to be fine

**Comment by pjakovits** [comment]
Have been using Ubuntu 16.04 OpenStack VM to host different Winery and other docker containers.

**Comment by Vogel612** [comment]
the permission issues are to be expected, since winery modifies the `.git` directory in the mounted volume as the docker container user.  
This automatically changes the owner of the affected files inside `.git` to `root:root`, which in turn makes them unreadable to git if it's executed from a "normal" user. 

To my knowledge there's no useful workaround to that, running `chown [user]:[user] -R .git` outside the container should reenable git functionality.

**Comment by avanhoorn** [comment]
> the permission issues are to be expected, since winery modifies the `.git` directory in the mounted volume as the docker container user.
> This automatically changes the owner of the affected files inside `.git` to `root:root`, which in turn makes them unreadable to git if it's executed from a "normal" user.
> 
> To my knowledge there's no useful workaround to that, running `chown [user]:[user] -R .git` outside the container should reenable git functionality.

Since I ran into the same permission issues (and didn't want to rebuild the container each time to hard-code the user ID), I use the following script: 

```
#!/usr/bin/env bash

GMT_DOCKER_COMPOSE="$(dirname $0)/deploy/compose/docker-compose-localparticles.yml"

PARTICLES_DIR="/path/to/your/radon-particles/"
DOCKER_COMPOSE_PARAMS="-f ${GMT_DOCKER_COMPOSE} -p radon-gmt"

chmod -R a+rwx ${PARTICLES_DIR}
docker-compose $DOCKER_COMPOSE_PARAMS pull
docker-compose $DOCKER_COMPOSE_PARAMS up
docker-compose $DOCKER_COMPOSE_PARAMS rm -f -s
sudo chown -R [YOUR-USER]:[YOUR-USER] ${PARTICLES_DIR}
sudo chmod -R 644 ${PARTICLES_DIR}
sudo chmod +X $(sudo find ${PARTICLES_DIR} -type d)

```

**Comment by miwurster** [comment]
@avanhoorn @pjakovits Could you guys please check if the image `miwurster/winery:latest` with the `-u` option solves the issue?

Try to run:

```
docker run -it -u `id -u` -p 8080:8080 \
  -e PUBLIC_HOSTNAME=localhost \
  -e WINERY_FEATURE_RADON=true \
  -e WINERY_REPOSITORY_PROVIDER=yaml \
  -v <path_on_your_host>:/var/repository \
  miwurster/winery:latest
```

Just replace `<path_on_your_host>` with the repo path on your host system.

**Comment by pjakovits** [comment]
> @avanhoorn @pjakovits Could you guys please check if the image `miwurster/winery:latest` with the `-u` option solves the issue?
>
> Try to run:
> ```
> docker run -it -u `id -u` -p 8080:8080 \
>   -e PUBLIC_HOSTNAME=localhost \
>   -e WINERY_FEATURE_RADON=true \
>   -e WINERY_REPOSITORY_PROVIDER=yaml \
>   -v <path_on_your_host>:/var/repository \
>   miwurster/winery:latest
> ```

I get errors about permissions:
```
mkdir: cannot create directory ‘//.winery’: Permission denied
2020/05/27 05:40:25 unable to create open //.winery/winery.yml: no such file or directory

java.io.IOException: Unable to create the directory [/usr/local/tomcat/webapps/winery-workflowmodeler]
```

**Comment by miwurster** [comment]
I think I fixed the permission issue. Could you briefly try it again?

**Comment by pjakovits** [comment]
Seems to work without issues now. No errors while launching container, and was able to create a service template, modify it and download CSAR. 

Just for information: These are the resulting file permissions outside the Docker container for the created service template files:
```
ubuntu@pellewinery:~/radon-particles$ ls -al servicetemplates/example.org.tosca.servicetemplates/testing3333333/
total 12
drwxr-x--- 2 ubuntu root   4096 May 27 10:25 .
drwxrwxr-x 4 ubuntu ubuntu 4096 May 27 10:25 ..
-rw-r----- 1 ubuntu root    459 May 27 10:25 ServiceTemplate.tosca
```

**Comment by miwurster** [comment]
Thanks @pjakovits for validation.

File: servicetemplates/example.org.tosca.servicetemplates/DataPipelineExample/ServiceTemplate.tosca [hunk 5]
```diff
@@ -0,0 +1,110 @@
+tosca_definitions_version: tosca_simple_yaml_1_3
```
Comment (miwurster) [inline servicetemplates/example.org.tosca.servicetemplates/DataPipelineExample/ServiceTemplate.tosca]: don't you want to change the namespace to `radon.blueprints.examples` or something more `radon` sepcific?

**Review by miwurster — CHANGES_REQUESTED** [review]
Would you like to push the changes from the `POST touch` command by yourself or should I do it after the merge?

File: servicetemplates/example.org.tosca.servicetemplates/DataPipelineExample/ServiceTemplate.tosca (hunk 5 above)
Comment (pjakovits) [reply servicetemplates/example.org.tosca.servicetemplates/DataPipelineExample/ServiceTemplate.tosca]: Good idea, changed it into radon.blueprints.examples

Comment (pjakovits) [reply servicetemplates/example.org.tosca.servicetemplates/DataPipelineExample/ServiceTemplate.tosca]: Also pushed changes after POST touch

**Review by miwurster — APPROVED** [review]

**Pull request merged**